core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -105,14 +100,13 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                     switch (currentFieldName) {
                     case "must":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            mustClauses.add(query);
-                        }
+                        mustClauses.add(query);
                         break;
                     case "should":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            shouldClauses.add(query);
+                        shouldClauses.add(query);
+                        // EmptyQueryBuilder does not add lucene query later, skip setting minuminShouldMatch
+                        if (query != EmptyQueryBuilder.PROTOTYPE) {
```

Comment:
So should I leave the check in place for now? Can't really move it to `toQuery` as long as we have no filter context there. If I leave it, where would you suggest we track this? I'd prefer a separate issue because it's somewhat unrelated to this but complicated enough to leave some of the writup I did above there so we don't forget. Thoughts?